### PR TITLE
Replace dateLib object with DateLib class

### LINKED ISSUE #2509: Inconsistent weekday headers and grid dates when using date-fns with default locale
## Code 

```tsx
import { DayPicker } from 'react-day-picker';
import { enAU } from 'date-fns/locale';
import * as dateFns from 'date-fns';

dateFns.setDefaultOptions({ locale: enAU });

export function App() {
  return <DayPicker dateLib={dateFns} mode='single' />;
}
```

## Expected Behavior

I would expect the weekday column headers to correctly match up with the dates in the grid as pictured below:

> Note: Friday 4th October 2024 is a correct date and appears in the two grids

locale explicitly set to `enAU` (`locale={enAU}`):
CleanShot 2024-10-04 at 14 48 48

locale explicitly set to `enUS` (`locale={enUS}`):
CleanShot 2024-10-04 at 14 50 18

## Actual Behavior

When `dateFns` is supplied as the `dateLib` and has been setup with the default locale of `enAU`, the weekday headers render as `enUS`, but the grid renders as `enAU`, resulting in incorrect weekdays lining up with dates:

> Note: Friday 5th October 2024 is **not** a correct date

CleanShot 2024-10-04 at 14 47 35

### PR DESCRIPTION
**Pull request by daveallie**
## What's Changed

- `dateLib` was an object containing a subset of date-fns functions, it is now a full class, which takes that subset and locale/formatting options as constructor arguments.
  - Although `DateLib` can take an undefined Locale, functions within `DateLib` which different return values based on locale, will require the DateLib class to have been initialised with a locale in order to be called.
    - This should allow for flexibility, while also enforcing the use of locale when it's needed.
  - Other formatting options like `weekStartsOn`, `firstWeekContainsDate`, `useAdditionalWeekYearTokens`, `useAdditionalDayOfYearTokens` are all passed to `DateLib` on initialization and explicitly not allowed to be passed in directly when calling those functions, instead we use the values that DateLib was initialized with
    - This effectively creates an instance of date-fns with a fixed set of formatting options
- I have tried to keep the APIs unchanged, across externally facing functions, using the optional `dateLib` as it was used before
  - `dateLib.options` exposes the formatting options it was initialized with, allowing for them to be passed down to user-provided label/format functions.

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation update

## Additional Notes

Fixes #2509
Related #2523
Related #2511

### CODE REVIEW

**Review by gpbl — COMMENTED** [review]
Great work here, @daveallie! I’ll need some time to review it. Thanks...

File: src/lib/dateLib.ts [hunk 1]
```diff
@@ -173,3 +178,206 @@ export const dateLib = {
...
 };
+
+/**
+ * The date library used by DayPicker. It's a subset of the date-fns functions
+ * with an explicit locale / formatting options.
+ *
+ * @private
+ * @internal
+ */
+export class DateLib {
```
Comment (gpbl) [inline src/lib/dateLib.ts:189]: Would this work somehow?

```suggestion
export class DateLib extends DefaultDateLib {
```

File: src/lib/dateLib.ts [hunk 2]
```diff
@@ -173,3 +178,206 @@ export const dateLib = {
...
+   */
+  constructor(
+    readonly options: Options = {},
+    dateLibBaseOverrides: Partial<DateLibBase> = {}
+  ) {
+    this.dateLibBase = {
+      ...defaultDateLibBase,
+      ...dateLibBaseOverrides
+    };
+  }
```
Comment (gpbl) [inline src/lib/dateLib.ts:204]: Is the `dateLibBase` property necessary? What about a `DateLib` class to extends a `DefaultDateLib`: Then, in the constructor:

```ts
if ('addWeeks' in dateLibBaseOverrides) this.addWeeks = dateLibBaseOverrides.addWeeks;
```

There might be reasons why we’re not doing this, but I can’t identify any just by looking at the code.

File: src/lib/dateLib.ts [hunk 3]
```diff
@@ -173,3 +178,206 @@ export const dateLib = {
...
+  /** Formats the given date using the specified format string. */
+  format = (date: DateArg<Date> & {}, formatStr: string) => {
+    if (!this.locale) {
+      throw new Error(
+        "DateLib needs to be initialized with locale to use the `format` function."
+      );
+    }
+
+    return format(date, formatStr, this.options);
+  };
```
Comment (gpbl) [inline src/lib/dateLib.ts:348]: The assumption here is that `this.options` includes `this.locale`, right? If so, why is `this.locale` needed separately? Couldn't we just:

```suggestion
    if (!this.options.locale) {
      throw new Error(
        "DateLib needs to be initialized with locale to use the `format` function."
      );
    }

    return format(date, formatStr, this.options);
  };
```

P.S. I love how well the code is documented. Thanks!

File: src/lib/dateLib.ts [hunk 4]
```diff
@@ -173,3 +178,206 @@ export const dateLib = {
...
+
+  /** Returns the week number for the given date. */
+  getWeek = (date: DateArg<Date> & {}) => {
+    if (!this.locale) {
+      throw new Error(
+        "DateLib needs to be initialized with locale to use the `getWeek` function."
+      );
+    }
+
```
Comment (gpbl) [inline src/lib/dateLib.ts:357]: Also, can’t we rely on a default `this.options.locale` to avoid these checks? 🤔

**Review by gpbl — COMMENTED** [review]
I find your solution very interesting, and there's definitely something to learn from your changes. I'm not very up-to-date with OOP in JavaScript, so here are some initial thoughts on `DateLib`.

The thing is, we should also test it, and I'm trying to find a way to simplify the code.

File: src/types/props.ts [hunk 5]
```diff
@@ -437,7 +437,7 @@ export interface PropsBase {
    * @since 9.0.0
    * @experimental
    */
-  dateLib?: Partial<DateLib> | undefined;
+  dateLib?: Partial<DateLibBase> | undefined;
```
[the remark this answers is not in the document]
Comment (daveallie) [reply src/types/props.ts:440]: yeah fair, I can leave the name unchanged for exporting, and adjust the internal name from `DateLib` to `DateLibImpl` or something

File: src/lib/dateLib.ts (hunk 4 above)
Comment (daveallie) [reply src/lib/dateLib.ts:357]: Default locale is what caused the original bug, by making locale optional on deeper calls in the library, we're allowing callers to forget to provide the known locale. For this implementation, I took the approach of still allowing locale optionally, but ensuring it's provided to use any function which is dependent on locale. This should help catch issues where locale is forgotten about in DateLib construction, but required when rendering.

Ultimately DateLib construction only happens inside `react-day-picker`, so the only time this is hit will be if we've written a bug, it's a footgun prevention mechanism

File: src/lib/dateLib.ts (hunk 3 above)
Comment (daveallie) [reply src/lib/dateLib.ts:348]: yeah absolutely! I think what you're suggesting is more clear, I was just using `this.locale` as shorthand, but it's more indirection.

File: src/lib/dateLib.ts [hunk 6]
```diff
@@ -173,3 +178,206 @@ export const dateLib = {
...
+      ...dateLibBaseOverrides
+    };
+  }
+
+  /**
+   * Creates a new instance of the date library with the default locale.
+   *
+   * @param options The formatting options for the date library.
+   */
+  static fromOptionsDefaultLocale(
```
[the remark this answers is not in the document]
Comment (daveallie) [reply src/lib/dateLib.ts:211]: yep, let me take another look at my approach, will attempt to rework this part

File: src/lib/dateLib.ts (hunk 2 above)
Comment (daveallie) [reply src/lib/dateLib.ts:204]: currently I've done it this way, as for `endOfWeek`, `format`, `getWeek`, and `startOfWeek` (locale required functions), I still want access to the original function with the correct name, but the interface from the class to not include the locale and have it pass in the internally known locale to the original function. I'll play around with another way to achieve the same thing a little cleaner.

**Comment by daveallie** [comment]
Thanks for the review @gpbl, I'm not an expert with OOP in JS either tbh, but it's pretty powerful if done right and this seems like a good use for it (wrapping up context and logic into a single instance). I had to wrap my head around the default args in the renderers which were a bit squirrelly, but this approach functions fine and provides a little more robust-ness for future expansion and prevention of issues in the future.

I'll throw up another revision in another few days addressing your feedback, if you spot anything else in there you'd like changed, please call it out. 🫡

**Comment by gpbl** [comment]
@daveallie, I had the chance to review the `DateLib` class in https://github.com/gpbl/react-day-picker/pull/2550.

You can check out the code here: [dateLib.ts](https://github.com/gpbl/react-day-picker/blob/gpbl/datelib/src/lib/dateLib.ts).

I still need to add tests, particularly for the weekday names that you met before, but this version is closer to the concept I had in mind.

If you have some time, I’d appreciate a quick review!

**Comment by gpbl** [comment]
PS. I tried configuring my PR to merge my changes into this one for an easier diff, but for some reason, I couldn't get it to work.

**Comment by gpbl** [comment]
@daveallie thanks so much for your extensive work here! I've merged #2550 which includes [the test](https://github.com/gpbl/react-day-picker/blob/8e396186e27b64452660003be6011055881994c3/examples/Testcase2511.tsx) for #2511.

**Comment by daveallie** [comment]
Thanks for picking this up and following it through @gpbl, appreciate it 🤝

**Pull request closed without merging**